### DISCUSSION TITLE
How to accept a Python string or Python `None` in a function

### DISCUSSION THREAD

**Discussion by hwittenborn** [PyO3/pyo3#2932] [Questions] [answered]
I've got a function that accepts two strings, but I'm wanting them to also be able to be set to `None`. Originally I thought of using this:

```rust
fn my_function(arg1: Option<String>, arg2: Option<String>) {
```

But this seems to make the arguments able to be omitted. I don't want the arguments to be able to be omitted though, I just want them to be nullable with `None`.

Is there any way to do what I'm wanting?

**Comment by adamreichold** [comment, marked as answer]
Try adding a signature indicating that the arguments are mandatory and do not have a default value, e.g.

```rust
#[pyfunction]
#[pyo3(signature=(arg1, arg2))]
fn my_function(arg1: Option<String>, arg2: Option<String>) {
```

**Comment by hwittenborn** [comment]
That doesn't seem to be working, I'm thinking it might be the context I'm using it in, I'm using it in a class constructor like so:

```rust
#[pyclass]
pub struct DistroArchPair {
    pub distro: Option<String>,
    pub arch: Option<String>
}

#[pymethods]
impl DistroArchPair {
    #[new]
    fn new(distro: Option<String>, arch: Option<String>) -> Self {
        Self {
            distro,
            arch
        }
    }
}
```

Adding `#[pyo3(signature=(arg1, arg2))]` below `#[new]` prints the following compiler error:

```
error: expected literal
  --> src/python.rs:25:22
   |
25 |     #[pyo3(signature=(arg1, arg2))]
   |                      ^
```
Sorry for being vague about it before, I just didn't want to send anything that complicated, but then I checked my code again and realized the affected part is quite tiny.

**Reply by birkenfeld** [reply]
Are you using PyO3 0.18 or earlier? This syntax is new in this version.

**Reply by adamreichold** [reply]
I will check, but I think for `#[new]` the signature must be put on the `#[pyclass]` instead?

**Reply by adamreichold** [reply]
Ah sorry, that was `text_signature`, `signature` should work, c.f. https://github.com/PyO3/pyo3/issues/2866 and https://pyo3.rs/v0.18.0/class#method-arguments

**Comment by hwittenborn** [comment]
> Are you using PyO3 0.18 or earlier? This syntax is new in this version.

That was it, everything worked fine after upgrading my version. Thanks a ton!

**Reply by hwittenborn** [reply]
Oh, thanks for the really quick support from all of you too! It's really helping me to get everything I need done :)

**Comment by davidhewitt** [comment]
For what it's worth, I'm adding a little bit of documentation for this case in #2934
